kong/runloop/handler.lua

```diff
@@ -1278,21 +1281,25 @@ return {
         var.upstream_connection = "keep-alive"
       end
 
+      local remote_addr = var.remote_addr
+      ctx.upstream_x_real_ip = remote_addr
+
       -- X-Forwarded-* Headers
+      -- Only store them in ngx.ctx and actually set to upstream header in access.after
+      -- so that plugins that terminates the requests doesn't got those headers
       local http_x_forwarded_for = var.http_x_forwarded_for
       if http_x_forwarded_for then
-        var.upstream_x_forwarded_for = http_x_forwarded_for .. ", " ..
-                                       realip_remote_addr
-
+        ctx.upstream_x_forwarded_for = http_x_forwarded_for .. ", " ..
+                                        realip_remote_addr
       else
-        var.upstream_x_forwarded_for = var.remote_addr
+        ctx.upstream_x_forwarded_for = remote_addr
```

Comment:
what is the difference between var.upstream_x_forwarded_for and ctx.upstream_x_forwarded_for?

Reply:
@Suika-Kong is working on some C land work for ngx.req.get_header, which may share some code with this one. Let's see if we can make both of them 3.0.

Reply:
> @Suika-Kong is working on some C land work for ngx.req.get_header, which may share some code with this one. Let's see if we can make both of them 3.0.

Should we discuss on a larger scope, like improving PDK's performance and design? People suggest to put all the request data in Lua land, and only submit them at once when all modifications to the Lua data is done.